File: Utils/src/hkccpacmanrobot/utils/map/ObstacleMap.java
package hkccpacmanrobot.utils.map;

import java.io.Serializable;
import java.util.HashMap;

/**
 * Created by 13058456a on 3/17/2015.
 */
public class ObstacleMap extends HashMap<MapKey, MapContent> implements Cloneable ,Serializable{
    public boolean isExist(MapUnit target) {
        return keySet().contains(target.key);
    }

    public void put(MapUnit mapUnit) {
        put(mapUnit.key, mapUnit.value);
    }

    public void get(MapUnit mapUnit) {
        mapUnit.value.set(get(mapUnit.key));
    }

    @Override
    public Object clone() {
        ObstacleMap newInstance = new ObstacleMap();
        for (MapKey key : this.keySet())
            newInstance.put((MapKey) key.clone(), (MapContent) get(key).clone());
        return newInstance;
    }
}

File: Utils/src/hkccpacmanrobot/utils/map/ObstacleMap.scala
package hkccpacmanrobot.utils.map

/**
 * Created by beenotung on 3/21/15.
 */

import scala.collection.mutable

class ObstacleMap extends mutable.HashMap[MapKey, MapContent] with Cloneable {
  def isExist(target: MapUnit): Boolean = {
    keySet.contains(target.key)
  }

  def put(mapUnit: MapUnit) {
    put(mapUnit.key, mapUnit.value)
  }

  def get(mapUnit: MapUnit) {
    mapUnit.value.set(get(mapUnit.key).get)
  }

  override def clone: AnyRef = {
    val newInstance: ObstacleMap = new ObstacleMap
    for (key <- keys) newInstance.put(key.clone.asInstanceOf[MapKey], get(key).clone.asInstanceOf[MapContent])
    return newInstance
  }
}